Repository/ContainerRepositoryFactory.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+/*
+ * This file is part of the Doctrine Bundle
+ *
+ * The code was originally distributed inside the Symfony framework.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ * (c) Doctrine Project, Benjamin Eberlei <kontakt@beberlei.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Doctrine\Bundle\DoctrineBundle\Repository;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\EntityRepository;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\ORM\Repository\RepositoryFactory;
+use Psr\Container\ContainerInterface;
+use Doctrine\Bundle\DoctrineBundle\DependencyInjection\Compiler\ServiceRepositoryCompilerPass;
+use Symfony\Component\DependencyInjection\Compiler\ServiceLocatorTagPass;
+
+/**
+ * Fetches repositories from the container or falls back to normal creation.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+final class ContainerRepositoryFactory implements RepositoryFactory
+{
+    private $managedRepositories = [];
+
+    private $container;
+
+    /**
+     * @param ContainerInterface $container A service locator containing the repositories
+     */
+    public function __construct(ContainerInterface $container = null)
+    {
+        // When DoctrineBundle requires Symfony 3.3+, this can be removed
+        // and the $container argument can become required.
+        if (null === $container && class_exists(ServiceLocatorTagPass::class)) {
+            throw new \InvalidArgumentException(sprintf('The first argument of %s::__construct() is required on Symfony 3.3 or higher.', self::class));
+        }
+
+        $this->container = $container;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRepository(EntityManagerInterface $entityManager, $entityName)
+    {
+        $metadata = $entityManager->getClassMetadata($entityName);
+        $repositoryServiceId = $metadata->customRepositoryClassName;
```

Comment:
unless I read something wrong, this variable doesn't seem to be used anywhere within the body of this method

Reply:
https://github.com/doctrine/DoctrineBundle/pull/727/files/4efcb135344e55403a0fd193c23a8c7249cfd7d0#diff-f36aa436e9aae3ea54bbaca0caf377e3R65 ;)

Reply:
10 years later 😆I agree with @Taluu , the variable doesn't seem to be used, as it used only used for the message of the `RuntimeException` that is thrown a few lines below.

But the `$customRepositoryName` class contains exactly the same value, and could be used instead. Maybe @weaverryan can tell more about it :)